src/Storages/MergeTree/MergeTreeIndexHypothesisMergedCondition.cpp

```diff
@@ -16,199 +21,371 @@ namespace ErrorCodes
+        : context(std::move(context_))
+        , table_expression(query_info.table_expression)
     {
-        active_atomic_formulas.insert(
-            std::end(active_atomic_formulas),
-            std::begin(hypothesis),
-            std::end(hypothesis));
+        auto ast_filter_node = buildFilterNode(query_info.query);
+
+        if (ast_filter_node)
+            expression_cnf = std::make_unique<CNF>(convertToCNF<is_ast>(ast_filter_node, context, table_expression));
+        else
+            expression_cnf = std::make_unique<CNF>(typename CNF::AndGroup{});
+
+        auto atomic_constraints_data = [&]
+        {
+            if constexpr (is_ast)
+                return constraints.getAtomicConstraintData();
+            else
+                return constraints.getQueryTreeData(context, table_expression).getAtomicConstraintData();
+        }();
+
+        for (const auto & atomic_formula : atomic_constraints_data)
+        {
+            auto atom = cloneAtom(atomic_formula, context);
+
+            if constexpr (is_ast)
+                atomic_constraints.push_back(std::move(atom.ast));
+            else
+                atomic_constraints.push_back(std::move(atom.node_with_hash.node));
+        }
     }
 
-    /// transform active formulas
-    for (auto & formula : active_atomic_formulas)
+    void addIndex(const MergeTreeIndexHypothesis & hypothesis_index) override
     {
-        formula = formula->clone(); /// do all operations with copy
-        auto * func = formula->as<ASTFunction>();
-        if (func && func->name == "less")
-            func->name = "lessOrEquals";
-        if (func && func->name == "greater")
-            func->name = "greaterOrEquals";
-    }
+        static const NameSet relations = { "equals", "notEquals", "less", "lessOrEquals", "greaterOrEquals", "greater"};
 
-    const auto weak_graph = std::make_unique<ComparisonGraph<ASTPtr>>(active_atomic_formulas);
+        // TODO: move to index hypothesis
+        std::vector<Node> compare_hypotheses_data;
+        std::vector<typename CNF::OrGroup> hypotheses_data;
+        auto cnf = convertToCNF<is_ast>(hypothesis_index.index.expression_list_ast->children.front(), context, table_expression);
 
-    bool useless = true;
-    expression_cnf->iterateGroups(
-        [&](const CNFQuery::OrGroup & or_group)
+        if constexpr (is_ast)
+            cnf.pullNotOutFunctions();
+        else
+            cnf.pullNotOutFunctions(context);
+
+        for (const auto & group : cnf.getStatements())
         {
-            for (const auto & atomic_formula : or_group)
+            if (group.size() == 1)
             {
-                CNFQuery::AtomicFormula atom{atomic_formula.negative, atomic_formula.ast->clone()};
-                pushNotIn(atom);
+                hypotheses_data.push_back(group);
+                auto atomic_formula = *group.begin();
+                auto atom = cloneAtom(atomic_formula, context);
+
+                assert(!atom.negative);
```

Comment:
you need to call `pushNotIn` before that, am I right?

Reply:
yes, but it's done inside `cloneAtom`.

Reply:
Maybe then we should rename it to `cloneAtomAndPushNot`?